firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/SetMutation.java

```diff
@@ -72,40 +71,32 @@ public String toString() {
   }
 
   @Override
-  public MaybeDocument applyToRemoteDocument(
-      @Nullable MaybeDocument maybeDoc, MutationResult mutationResult) {
-    verifyKeyMatches(maybeDoc);
+  public void applyToRemoteDocument(Document document, MutationResult mutationResult) {
+    verifyKeyMatches(document);
 
     // Unlike applyToLocalView, if we're applying a mutation to a remote document the server has
     // accepted the mutation so the precondition must have held.
-
-    SnapshotVersion version = mutationResult.getVersion();
-
-    ObjectValue newData = value;
     if (mutationResult.getTransformResults() != null) {
-      List<Value> transformResults =
-          serverTransformResults(maybeDoc, mutationResult.getTransformResults());
-      newData = transformObject(newData, transformResults);
+      Map<FieldPath, Value> transformResults =
+          serverTransformResults(document, mutationResult.getTransformResults());
+      ObjectValue localValue = value.clone();
+      localValue.set(transformResults);
+      document.asFoundDocument(mutationResult.getVersion(), localValue).withCommittedMutations();
+    } else {
+      document.asFoundDocument(mutationResult.getVersion(), value.clone()).withCommittedMutations();
```

Comment:
You could avoid duplicating most of the arguments to `asFoundDocument` by following the old form of this method:

```java
ObjectValue newData = value.clone();
if (mutationResult.getTransformResults() != null) {
  transformResults = ...;
  newData.set(transformResults);
}

return document.asFoundDcument(mutationResult.getVersion(), newData).withCommittedMutations();
```

Reply:
Done (but some of this is no longer applicable)